File: 01-04-2025_DFS_of_Graph.java
import java.util.*;

class Solution {
    public void dfs(int i, List<List<Integer>> adj, List<Integer> ans, Map<Integer, Boolean> visited) {
        ans.add(i);
        visited.put(i, true);
        for (int z : adj.get(i)) {
            if (!visited.getOrDefault(z, false)) {
                dfs(z, adj, ans, visited);
            }
        }
    }

    public List<Integer> dfs(List<List<Integer>> adj) {
        List<Integer> ans = new ArrayList<>();
        Map<Integer, Boolean> visited = new HashMap<>();
        dfs(0, adj, ans, visited);
        return ans;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int tc = scanner.nextInt();
        while (tc-- > 0) {
            int V = scanner.nextInt();
            scanner.nextLine(); // Consume the newline
            List<List<Integer>> adj = new ArrayList<>();

            for (int i = 0; i < V; i++) {
                String line = scanner.nextLine();
                List<Integer> node = new ArrayList<>();
                if (!line.isEmpty()) {
                    for (String num : line.split(" ")) {
                        node.add(Integer.parseInt(num));
                    }
                }
                adj.add(node);
            }

            Solution obj = new Solution();
            List<Integer> ans = obj.dfs(adj);
            for (int num : ans) {
                System.out.print(num + " ");
            }
            System.out.println();
        }
        scanner.close();
    }
}

File: 01-04-2025_DFS_of_Graph.py
class Solution:
    def dfs_util(self, i, adj, ans, visited):
        ans.append(i)
        visited[i] = True
        for z in adj[i]:
            if not visited[z]:
                self.dfs_util(z, adj, ans, visited)

    def dfs(self, adj):
        ans = []
        visited = {i: False for i in range(len(adj))}
        self.dfs_util(0, adj, ans, visited)
        return ans

# Driver code
if __name__ == "__main__":
    tc = int(input().strip())
    for _ in range(tc):
        V = int(input().strip())
        adj = []
        
        for _ in range(V):
            line = input().strip()
            if line:
                node = list(map(int, line.split()))
            else:
                node = []
            adj.append(node)
        
        obj = Solution()
        ans = obj.dfs(adj)
        print(" ".join(map(str, ans)))
